Packages/com.unity.inputsystem/InputSystem/Devices/ScreenKeyboard.cs

```diff
@@ -0,0 +1,298 @@
+    /// </summary>
+    /// Note: We use delegates here instead of events to disallow a potential bug from users
+    ///       If these would be events, the order in which callbacks would invoked would be undefined.
+    ///       Imagine if in one of the inputFieldTextChanged callbacks we change the input field text.
+    ///       The other inputFieldTextChanged would still receive the original changed text which is wrong and could lead to potential problems.
+    /// Note: We use this approach with class callbacks, since it's more future proof, if we add an additional callback in the future, it won't break existing user projects.
+    internal class ScreenKeyboardCallbacks
+    {
+        /// <summary>
+        /// Is fired whenever screen keyboard state changes
+        /// </summary>
+        internal Action<ScreenKeyboardState> stateChanged { set; get; }
+
+        /// <summary>
+        /// Is fired whenever input field text changes.
+        /// It doesn't matter if input field is hidden or not, this callback would still be invoked
+        /// </summary>
+        internal Action<string> inputFieldTextChanged { set; get; }
+
+        /// <summary>
+        /// Is fired whenever input field text selection changes.
+        /// </summary>
+        internal Action<RangeInt> inputFieldSelectionChanged { set; get; }
+    }
+
+    /// <summary>
+    /// ScreenKeyboard base class for platform specific implementation.
+    ///
+    /// Known issues:
+    /// * There's a screen keyboard behavior differences in regards to input and focus loss:
+    ///   - [Android] When screen keyboard is shown, the main Unity window stops receiving input events, when clicking on Unity window, screen keyboard looses focus and closes itself
+    ///     (While it's possible to keep screen keyboard opened on focus loss, currently there's no way to make Unity window keep receiving input events like touch while screen keyboard is shown,
+    ///      this matches the behavior in old TouchScreenKeyboard implementation)
+    ///   - [iOS] When screen keyboard is shown, the main Unity window will continue receiving input events, when clicking on Unity window, screen keyboard will continue to be shown
+    ///   - [WSA] When screen keyboard is shown, the main Unity window will continue receiving input events, when clicking on Unity window, screen keyboard will continue to be shown
+    /// </summary>
+    internal abstract class ScreenKeyboard : IDisposable
+    {
+        protected ScreenKeyboardState m_KeyboardState;
+        protected ScreenKeyboardShowParams m_ShowParams;
+        private ScreenKeyboardCallbacks m_Callbacks;
+
+        /// <summary>
+        /// Returns the state of the keyboard
+        /// </summary>
+        public ScreenKeyboardState state
+        {
+            get => m_KeyboardState;
+        }
+
+        /// <summary>
+        /// Cleanup any platform specific resources
+        /// </summary>
+        public virtual void Dispose()
```

Comment:
Looks like every single implementation is expected to override this in order to at least null the static instance?
If such expectation is correct, then this better be abstract.

Reply:
Actually, we do we even have per-class static instance and one more global instance?

Reply:
no, that shouldn't be possible, only one static instance of specific keyboard should be ever alive

Reply:
I think static instance might not be even needed anymore, i'll check

Reply:
Removed static instances

Reply:
Now that you have removed statics from implementation, it makes sense for base class to have do-nothing Dispose(), instead of each implementation overriding it and only Android one doing something :)

Reply:
so it's still makes sense to have virtual in base class, otherwise Android Dispose won't get called

Reply:
iOS has also dispose

Reply:
Fixed now